Deprecate public Controller properties

robertpustulka opened the pull request:
Refs #10946

I'm afraid it's hard to change the visibility of `$request` and `$response` properties now, so I kept them public but added deprecation annotations to them.

`$name`, `$plugin`, `$autoRender` properties are now protected and accessed via magic methods for BC.

It was also hard to refactor the code to use request controller name and plugin params as those not always share the same state, at least in tests. So these are still controller properties. We can get back to this in 4.0 though.

I added deprecation annotations to `$components` and `$helpers` as well because those are deprecated since the 3.0 I believe.

After this is reviewed and merged, I'll update the core to use new methods.

Commit:
Add setters and getters for name and plugin.

Commit:
Add setters and getters for request and response.

Commit:
Add setter and getter for autoRender flag.

Commit:
Add deprecations for $components and $helpers properties.

chinpei215 commented on src/Controller/Controller.php:
```diff
@@ -159,7 +163,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      * @var bool
      */
-    public $autoRender = true;
+    protected $autoRender = true;
```
I think this change is meaningless. Because when I touch this property, I am in the action of the controller usually. So I can continue to use this property without any warnings even if I am using some great IDE. Probably, it's the time to allow us to use `private` modifiers. To be honest, I think clean APIs wouldn't prefer setters/getters for public/protected properties or getters for private mutable objects.

ADmad replied:
This change is not BC and actually not meaningless. There could be some doing `$this->getController()->autoRender = false` in a component.

robertpustulka replied:
Subclasses could call `$this->enable/disableAutoRender()` in their initialize methods.

markstory replied:
They could. But making just _this_ property private would introduce inconsistencies with other parts of the framework and break compatibility.

inoas commented:
I'd either have also added $request and $response to the magic getters/setters or (preferably) not used magic getters/setters at all but just mark the properties with a deprecation warning that their visibility will change from public to private in 4.0 and that you should use the appropriate getter/setter methods.

robertpustulka commented:
> I'd either have also added $request and $response to the magic getters/setters

I tried, but many of the tests just failed. I'm afraid it's not that easy to do this.


> Because when I touch this property, I am in the action of the controller usually.

For me that is a reason this property shouldn't be public.

inoas commented on src/Controller/Controller.php:
```diff
@@ -124,6 +126,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      *
      * @var \Cake\Http\ServerRequest
      * @link https://book.cakephp.org/3.0/en/controllers/request-response.html#request
+     * @deprecated 3.5.0 Use getRequest()/setRequest instead.
```
This is one of the places where I would strongly prefer `request()` for getting as that is the 99% case and everything else is annoying verbosity. See https://github.com/cakephp/cakephp/issues/10930.

I am not even sure if the request object should be changed? Should it? Or should it merely be cloned via `withFoo()`?

markstory replied:
@inoas You wouldn't need to use the methods *inside* the controller actions. In that context you could use the protected properties.

ADmad commented:
Changing public properties to protected in not backwards compatible, please revert it. I showed an example above how it can create problems.

robertpustulka commented:
@ADmad There are magic `__get` and `__set` updated to handle this. Please look at the diff :)

robertpustulka commented on src/Controller/Controller.php:
```diff
@@ -124,6 +126,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      *
      * @var \Cake\Http\ServerRequest
      * @link https://book.cakephp.org/3.0/en/controllers/request-response.html#request
+     * @deprecated 3.5.0 Use getRequest()/setRequest instead.
```
@ionas I prefer the consistency. All getters here have a prefix, so this one should also have it. BTW `setRequest` has been around for the while and is used to set the state of the controller.

ADmad commented:
Sorry missed that :)

ADmad commented on src/Controller/Controller.php:
```diff
@@ -390,6 +418,85 @@ public function __set($name, $value)
+     * @param bool $autoRender Flag.
+     * @return $this
+     */
+    public function enableAutoRender($autoRender = true)
```
If this method has `enable` prefix then it should only enable auto rendering. Currently doing `enableAutoRender(false)` would disable auto rendering which is not good for readability.

robertpustulka replied:
I was following the examples in a query builder, like `enableAutoFields` and `enableHydration` which accept `false`.

ADmad replied:
So either the method should be name `setAutoRender()` or two separate methods `enableAutoRender()` and `disableAutoRender()` should be added which don't take any argument.

robertpustulka replied:
@ADmad are you suggesting that we should do the same in a query builder? That obviously makes more sense then calling `enableSomething(false)`

markstory replied:
I've found `enableThing(false)` to be clear enough, but I might be an outlier there.

ADmad replied:
For me personally even `something(false)` is good enough :)

chinpei215 commented:
What do you think about benefits of setters/getters? In my opinion, setters ensure to set unexpected values to the property. But if it is a `protected` property, I can write `$this->autoRender = 0;` in my controller for example. So this kind of setters/getters don't have any advantages actually.

robertpustulka commented:
@chinpei215 Right, I can add a typecheck to the `enableAutoRender` since we're not on PHP 7 yet.

chinpei215 commented:
Even if we are on PHP7, properties will be weakly typed. Are you planning to change the modifier to `private` in 4.0? Then I am ok. But if you are not planning that, I am not in favor of this change, as people would continue to write `$this->autoRender = true;`.

robertpustulka commented:
@chinpei215 I can't recall if there has been any discussion or decision made on `private` properties. But I also think that some of the properties could become private.

chinpei215 commented:
OK. Thank you for your reply :)

markstory commented on src/Controller/Controller.php:
```diff
@@ -410,6 +517,31 @@ public function setRequest(ServerRequest $request)
+     */
+    public function setResponse(Response $response)
+    {
+        $this->response = $response;
```
Because responses and requests are immutable I'm wondering if we should start thinking of ways to make working with responses less clunky. Using setResponse() is going to get annoying.

robertpustulka replied:
Are there many cases where a request is modified in a controller scope?

markstory replied:
I'm sure there are. People do things like mutate request data quite often.

robertpustulka replied:
Right, I forgot about data. 

In controller methods it will be still possible to do `$this->request = $this->request->withSomething()`

robertpustulka commented on src/Controller/Controller.php:
```diff
@@ -206,7 +212,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      * @var string
      */
-    public $plugin;
+    protected $plugin;
```
We could but I've seen some tests that set the plugin property. I'm not sure how this would work with immutable requests but this could be a matter of refactoring tests.

markstory replied:
We can start by refactoring our tests.

saeideng commented:
I think 
1)we need change scope of `request` and `response` to protected
2) un-deprecate   accessing `this->request` and `$this->response`
3)set those by `__set()` and validate if received object is `serverRequest` and `response` Object or no
4)run a error if received object is invalid 
5)get by `__get()` 

so we able to use 
`$this->request = $this->request->withSomething()`
`$this->response  = $this->response->withSomething()`
safety

chinpei215 commented:
No. When you assign a value to a protected property from a derived class, `__set()` will not be called. So people can write `$this->request = 1;` unsafely. Please read the earlier discussion about `private` modifier.

saeideng commented:
what happen if we change `public $request` to `protected $_request` or `protected $_other_request_property`

saeideng commented:
i means
```
     public function setRequest(ServerRequest $request)
     {
         $this->request = $request;
         $this->plugin = $request->getParam('plugin') ?: null;
 
         if ($request->getParam('pass')) {
             $this->passedArgs = $request->getParam('pass');
         }

        return $this;
    }
```
```
         $deprecated = [
            'request' => 'setRequest',
            'response' => 'setResponse'
        ];
        if (isset($deprecated[$name])) {
            $method = $deprecated[$name];

            return $this->{$method}();
        }
```

saeideng commented:
you can not pass `1` to `public function setRequest(ServerRequest $request)`
I misunderstood ?

chinpei215 commented:
> I misunderstood ?

Yes. First, people can access the protected property directly: `$this->_request = 1;`.
Second, if you declare `public $request = 1;` in a derived class, the property is evaluated as `1` in any methods of the parent class **if `private $request` is not declared** . So, for example, `$this->referer()` will cause fatal error because it calls `$this->request->referer()` internally.

saeideng commented:
I forgot we are in subcluss of cor controller :smiley:

Commit:
Split enableAutoRender($flag) 
into enableAutoRender() and disableAutoRender().

Deprecated properties test added.

Commit:
Deprecate $plugin property in favor of $request->getParam('plugin').

robertpustulka commented:
I've added the `disableAutoRender()` method and marked the property deprecated. It may convince users to using new  methods in their actions instead.
I've also removed get/setPlugin methods, deprecated the property in favor of the requets param.

I'm going to remove the deprecated properties and method calls in another PR once this is merged.

Commit:
Deprecate $autoRender property.

robertpustulka commented:
Is there a need for deprecation warnings for the properties in __get and __set? I mean something like with the deprecated view builder properties.

cleptric commented:
@robertpustulka I would add them after 3.5 shipped.

markstory commented on src/Controller/Controller.php:
```diff
@@ -251,14 +252,15 @@ public function __construct(ServerRequest $request = null, Response $response =
 
         $this->setRequest($request !== null ? $request : new ServerRequest());
-        $this->response = $response !== null ? $response : new Response();
+        $this->setResponse($response !== null ? $response : new Response());
```
This could use `?:` now I think.

markstory commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -1145,15 +1131,57 @@ public function testAutoRender()
+
+        try {
+            $controller->$property = $value;
+//
```
Looks like a mistake.

robertpustulka committed:
Small fixes.

markstory commented:
I am going to hold off merging this until 3.6 development starts. 3.5 is already a big release.

dereuromark commented:
Looks like this is now pretty much good to go.

markstory commented on src/Controller/Controller.php:
```diff
@@ -206,6 +213,7 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
      * Automatically set to the name of a plugin.
      *
      * @var string
+     * @deprecated 3.5.0 Use `$this->request->getParam('plugin')` instead.
```
These numbers should be `3.6.0` now.

robertpustulka committed:
Updated deprecation version numbers. Added `@since` tags.

robertpustulka committed:
Added warnings for deprecated properties.

robertpustulka committed:
Fix tests

robertpustulka commented:
I've updated the numbers, added `@since` tags and fixed some tests to use getters and setters.

dereuromark commented on src/Controller/Controller.php:
```diff
@@ -159,8 +163,9 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
+     * @deprecated 3.6.0 Use enableAutoRender()/disableAutoRender() and isAutoRenderEnabled() instead.
      */
-    public $autoRender = true;
+    protected $autoRender = true;
```
Will this be a BC issue from components setting this on the controller?

robertpustulka replied:
With this it should be ok:
https://github.com/robertpustulka/cakephp/blob/e2d894e65b7731b971a8ded78d42971bb32b7c27/src/Controller/Controller.php#L403-L407

saeideng commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -1086,7 +1086,121 @@ public function testBeforeRenderViewVariables()
+        $message = false;
+
+        set_error_handler(function ($errno, $errstr) use (&$message) {
+            $message = ($errno === E_USER_DEPRECATED ? $errstr : false);
```
`$message` not used 
see my recent PR

robertpustulka replied:
I'll wait with updating tests until #11231 is merged if that's what you mean.

saeideng replied:
yes 
I'll update it tonight

robertpustulka committed:
Added getPlugin() and setPlugin() methods for easy plugin name access.

robertpustulka committed:
Added deprecation warnings.

robertpustulka committed:
Fix deprecated plugin access.

robertpustulka commented:
The branch is up to date with `3.next`.

I added `plugin` getter and setter to the controller. Some tests seem to expect that the state of controller plugin and request plugin parameters may differ and I didn't want to break that behavior by delegating controller plugin property to the request entirely.

This could be refactored in 4.x so the `name` and `plugin` parameters would be accessed only via request.

robertpustulka committed:
Remove deprecation notice from a protected property.

saeideng commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -1115,6 +1115,166 @@ public function testBeforeRenderViewVariables()
+     *
+     * @return array
+     */
+    public function deprecatedControllerPropertyProvider()
```
if structure of this and `deprecatedViewPropertyProvider` is same
you can reduce test by adding content of this to `deprecatedViewPropertyProvider` 
so you can remove  bellow test IMO
```php
testDeprecatedControllerPropertyGetterMessage
testDeprecatedControllerPropertySetterMessage
testDeprecatedControllerProperty
```
also `expectedExceptionMessageRegExp /^Controller::\$\w+ is deprecated(.*)/` is same for 9 deprecated methods and this RegEx just check first of message (until `'Controller::$%s is deprecated.` )not all

saeideng replied:
we no need to check all `'Controller::$%s is deprecated. Use $this->%s instead.'` message IMO

robertpustulka replied:
Deprecated view properties point to the `$controller->viewBuilder()->method()` so this is a bit different.

saeideng replied:
OK :+1:  Your are right 
I forget the `->viewBuilder()`

robertpustulka commented:
Can it be merged? I'd like to do the similar with `View` and `Cell` classes.

saeideng commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -1115,6 +1115,166 @@ public function testBeforeRenderViewVariables()
+     */
+    public function testDeprecatedControllerPropertySetterMessage($property, $getter, $setter, $value)
+    {
+        error_reporting(E_ALL);
```
now
you should use `withErrorReporting()` for wrapping   `error_reporting()`
and `deprecated()` for testing deprecated methods

in here you must use `deprecated()`

robertpustulka replied:
@saeideng tests have been updated

saeideng replied:
nice :+1:

ADmad commented on src/Controller/Controller.php:
```diff
@@ -386,17 +422,113 @@ public function __set($name, $value)
+     * @return string|null
+     * @since 3.6.0
+     */
+    public function getPlugin()
```
I don't think we should `getPlugin()`, `setPlugin()`. Just make people use `$this->request->getParam('plugin')` instead.

robertpustulka replied:
I tried this but some tests were failing. It looked to me that there were some cases where different state od request plugin param and controller plugin name was expected.

ADmad replied:
Okay, that's sad :disappointed:.

robertpustulka committed:
Update tests.

robertpustulka commented:
Can we merge this? I had to resolve conflicts like 3rd or 4th time already...

markstory commented on tests/TestCase/Controller/ControllerTest.php:
```diff
@@ -1085,6 +1085,169 @@ public function testBeforeRenderViewVariables()
+    {
+        $controller = new AnotherTestController();
+        $controller->{$setter}($value);
+        $this->withErrorReporting(E_ALL, function () use ($controller, $property) {
```
The default error level is `E_ALL` why do you need to increase it here?

robertpustulka replied:
Must be some out of date code. I'll fix that.

saeideng replied:
default yes 
but we can change default in future if needed 
and in this case test must return  error/warning , so `E_ALL` is true IMO

robertpustulka replied:
@markstory I think that @saeideng has a point here. What is your opinion on this?

markstory replied:
I would prefer to keep our test suite at `E_ALL` and disable deprecation warnings as needed via `deprecated()`.

saeideng replied:
so we no need to assert deprecation warnings in tests like other place
